pkg/rpcserver/rpc_server.go

```diff
@@ -0,0 +1,127 @@
+package rpcserver
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/signal"
+	"syscall"
+
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+
+	"github.com/stripe/stripe-cli/rpc"
+)
+
+// Config provides the configuration for the gRPC server
+type Config struct {
+	// Port is the port number to listen to on localhost
+	Port int
+
+	// Info, error, etc. logger. Unrelated to API request logs.
+	Log *log.Logger
+}
+
+// RPCServer is the gRPC server
+type RPCServer struct {
+	cfg *Config
+}
+
+// ConfigOutput is the server config written to stderr for clients to read
+type ConfigOutput struct {
+	// Address is the localhost address of the gRPC server
+	Address string `json:"address"`
+}
+
+// New creates a new gRPC server
+func New(cfg *Config) *RPCServer {
+	if cfg.Log == nil {
+		cfg.Log = &log.Logger{Out: ioutil.Discard}
+	}
+
+	return &RPCServer{
+		cfg: cfg,
+	}
+}
+
+func withSIGTERMCancel(ctx context.Context, onCancel func()) context.Context {
+	// Create a context that will be canceled when Ctrl+C is pressed
+	ctx, cancel := context.WithCancel(ctx)
+
+	interruptCh := make(chan os.Signal, 1)
+	signal.Notify(interruptCh, os.Interrupt, syscall.SIGTERM)
+
+	go func() {
+		<-interruptCh
+		onCancel()
+		cancel()
+	}()
+	return ctx
+}
+
+// Run starts the gRPC server
+func (srv *RPCServer) Run(ctx context.Context) error {
+	ctx = withSIGTERMCancel(ctx, func() {
```

Comment:
I think we need to think about this a bit more. That's how we handle cmd+c right now in other functions, but it's creating weird race conditions when 2 different commands are calling each others (such as Playback internally running Listen).

`withSIGTERMCancel` is purely a command-line specific UX and I'm wondering if this should be the one, single bit of additional logic we could allow in the cobra command.

Reply:
Agree, also we should probably DRY it